runtime/src/iree/hal/drivers/local_task/task_semaphore.c

```diff
@@ -369,6 +370,7 @@ iree_status_t iree_hal_task_semaphore_multi_wait(
         status = iree_hal_task_semaphore_acquire_timepoint(
             semaphore, semaphore_list->payload_values[i], timeout, timepoint);
         if (iree_status_is_ok(status)) {
+          timepoint->semaphore = &semaphore->base;
```

Comment:
I don't feel 100% comfortable with the fact that with this code, iiuc, `timepoint->semaphore`  is NULL initially but then permanently switches to non-NULL once `iree_hal_task_semaphore_multi_wait` has been called. This seems to be unintentially introducing a boolean "mode" here.

Would it be better instead to create a local array of semaphore pointers, local in this `iree_hal_task_semaphore_multi_wait` function? A small heap allocation looks like it would be small overhead compared to what this function already does, this would remove the "mode" issue and would also avoid inflating `iree_hal_task_timepoint_t` ?

Reply:
I think you're asking for what this already does :)
The iree_hal_task_timepoint_t is not cleared - it's only ever set by this code and then checked by it - only the inner iree_hal_timepoint_t is ever modified externally.

Reply:
(the NULL check at the cancel below is just because that's cleanup code and if acquire_timepoint fails midway through a list we only want to cleanup the ones that have been initialized)

Reply:
The timepoints are local to that function - they never escape - so there's nothing but this function that ever knows or cares*. Within the function the timepoint array is created, up to N are initialized, the wait is performed, all initialized timepoints are cancelled, and then the function returns.

* technically they escape into the semaphore timepoint list and may be touched by other threads while in there, but the acquire_timepoint is what inserts them and the cancel_timepoint is what erases them so there's no way for the function to exit without doing that (unless there's a bug) so they're still only extant within the scope of the function.

Reply:
specifically, here's the alloc of the list:
https://github.com/google/iree/blob/9d95e114bad481a0a3703bb81c15267ff1ac3366/runtime/src/iree/hal/drivers/local_task/task_semaphore.c#L357-L358

and the free:
https://github.com/google/iree/blob/9d95e114bad481a0a3703bb81c15267ff1ac3366/runtime/src/iree/hal/drivers/local_task/task_semaphore.c#L401

if a timepoint exists after that area reset there's bigger problems :)

Reply:
Yup, OK. Looks like we are talking the same thing, but you're focusing on what the code is today, while I'm thinking about it in terms of "could this data structure be misused in the future". Anyway, this is a tiny corner of the universe.